contracts/cw3-multiple-choice/src/state.rs

```diff
@@ -0,0 +1,172 @@
+use std::convert::TryInto;
+
+use cosmwasm_std::{Addr, BlockInfo, CosmosMsg, Empty, StdError, StdResult, Storage, Uint128};
+use cw3::Status;
+use cw_storage_plus::{Item, Map};
+use cw_utils::{Duration, Expiration};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use crate::{helpers::votes_needed, msg::Threshold};
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+pub struct Config {
+    pub threshold: Threshold,
+    pub max_voting_period: Duration,
+    pub proposal_deposit: Uint128,
+    pub refund_failed_proposals: Option<bool>,
+    pub gov_token_address: Addr,
+    pub parent_dao_contract_address: Addr,
+}
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+pub struct Proposal {
+    pub title: String,
+    pub description: String,
+    pub choices: Vec<String>,
+    pub proposer: Addr,
+    pub start_height: u64,
+    pub expires: Expiration,
+    pub msgs: Vec<Vec<CosmosMsg<Empty>>>,
+    pub status: Status,
+    // Selected choice is index in choices
+    pub selected_choice: Option<usize>,
+    /// Pass requirements
+    pub threshold: Threshold,
+    /// The total weight when the proposal started (used to calculate percentages)
+    pub total_weight: Uint128,
+    /// summary of existing votes
+    pub votes: Votes,
+    /// Amount of the native governance token required for voting
+    pub deposit: Uint128,
+}
+
+impl Proposal {
+    /// current_status is non-mutable and returns what the status should be.
+    /// (designed for queries)
+    pub fn current_status(&mut self, block: &BlockInfo) -> Status {
+        let mut status = self.status;
+
+        // if open, check if voting is passed or timed out
+        if status == Status::Open && self.is_passed(block) {
+            status = Status::Passed;
+            self.selected_choice = Some(
+                self.votes
+                    .vote_counts
+                    .iter()
+                    .enumerate()
+                    .max_by(|a, b| a.cmp(b))
+                    .map(|(idx, _)| idx)
+                    .unwrap(),
+            );
+        }
+        if status == Status::Open && self.expires.is_expired(block) {
+            status = Status::Rejected;
+        }
+
+        status
+    }
+
+    /// update_status sets the status of the proposal to current_status.
+    /// (designed for handler logic)
+    pub fn update_status(&mut self, block: &BlockInfo) {
+        self.status = self.current_status(block);
+    }
+
+    pub fn is_passed(&self, block: &BlockInfo) -> bool {
```

Comment:
We likely also want a `is_rejected` function here as well to allow proposals to be properly shown as rejected in the UI when there is no way they can pass. You should be able to tweak this one: https://github.com/DA0-DA0/dao-contracts/blob/5483427cfedb1ae2844a6e403369c7f3d601a6d1/contracts/cw-govmod-single/src/proposal.rs#L128